File: classifier/data/raw/ai/samples/project_euler_java/Problem084.java
import java.util.Arrays;

/** Project Euler 084: Monopoly odds with two four-sided dice. */
public class Problem084 {
    private static final int GO = 0, JAIL = 10, G2J = 30;
    private static final int[] CHANCE = {7, 22, 36};
    private static final int[] CHEST = {2, 17, 33};

    public static void main(String[] args) {
        // A state includes the current square and the current run of doubles.
        double[][] transition = new double[120][120];
        for (int square = 0; square < 40; square++) for (int doubles = 0; doubles < 3; doubles++) {
            int from = state(square, doubles);
            for (int a = 1; a <= 4; a++) for (int b = 1; b <= 4; b++) {
                double probability = 1.0 / 16.0;
                if (a == b && doubles == 2) {
                    transition[from][state(JAIL, 0)] += probability;
                } else {
                    int nextDoubles = a == b ? doubles + 1 : 0;
                    land(transition, from, (square + a + b) % 40, nextDoubles, probability);
                }
            }
        }

        double[] distribution = new double[120];
        distribution[state(GO, 0)] = 1;
        for (int iteration = 0; iteration < 10_000; iteration++) {
            double[] next = new double[120];
            for (int from = 0; from < 120; from++) if (distribution[from] != 0)
                for (int to = 0; to < 120; to++) next[to] += distribution[from] * transition[from][to];
            distribution = next;
        }
        double[] squares = new double[40];
        for (int square = 0; square < 40; square++)
            for (int doubles = 0; doubles < 3; doubles++) squares[square] += distribution[state(square, doubles)];
        Integer[] order = new Integer[40];
        for (int i = 0; i < 40; i++) order[i] = i;
        Arrays.sort(order, (x, y) -> Double.compare(squares[y], squares[x]));
        System.out.printf("%02d%02d%02d%n", order[0], order[1], order[2]);
    }

    private static int state(int square, int doubles) { return square * 3 + doubles; }

    private static void land(double[][] transition, int from, int square, int doubles, double probability) {
        if (square == G2J) { transition[from][state(JAIL, 0)] += probability; return; }
        if (contains(CHEST, square)) {
            // Community Chest: GO, JAIL, or one of fourteen cards which do not move us.
            transition[from][state(GO, doubles)] += probability / 16;
            transition[from][state(JAIL, 0)] += probability / 16;
            transition[from][state(square, doubles)] += probability * 14 / 16;
        } else if (contains(CHANCE, square)) {
            // The six non-moving Chance cards are combined into one branch.
            transition[from][state(square, doubles)] += probability * 6 / 16;
            moveCard(transition, from, GO, doubles, probability / 16);
            moveCard(transition, from, JAIL, doubles, probability / 16);
            moveCard(transition, from, 11, doubles, probability / 16); // C1
            moveCard(transition, from, 24, doubles, probability / 16); // E3
            moveCard(transition, from, 39, doubles, probability / 16); // H2
            moveCard(transition, from, 5, doubles, probability / 16);  // R1
            moveCard(transition, from, nextRailroad(square), doubles, probability * 2 / 16);
            moveCard(transition, from, nextUtility(square), doubles, probability / 16);
            land(transition, from, (square + 37) % 40, doubles, probability / 16); // back three
        } else {
            transition[from][state(square, doubles)] += probability;
        }
    }

    private static void moveCard(double[][] t, int from, int square, int doubles, double p) {
        t[from][state(square, square == JAIL ? 0 : doubles)] += p;
    }
    private static int nextRailroad(int square) {
        if (square == 7) return 15;
        if (square == 22) return 25;
        return 5;
    }
    private static int nextUtility(int square) { return square == 22 ? 28 : 12; }
    private static boolean contains(int[] values, int value) {
        for (int item : values) if (item == value) return true;
        return false;
    }
}

File: classifier/data/raw/ai/samples/project_euler_zig/problem_084.zig
const std = @import("std");

// Project Euler 84: modal squares for Monopoly with two four-sided dice.
const squares = 40;
const states = squares * 3;

fn addRaw(next: *[states]f64, position: usize, doubles: usize, probability: f64) void {
    next[position * 3 + doubles] += probability;
}

fn addLanding(next: *[states]f64, position: usize, doubles: usize, probability: f64) void {
    if (position == 30) {
        addRaw(next, 10, 0, probability);
    } else if (position == 2 or position == 17 or position == 33) {
        addRaw(next, 0, doubles, probability / 16.0);
        addRaw(next, 10, 0, probability / 16.0);
        addRaw(next, position, doubles, probability * 14.0 / 16.0);
    } else if (position == 7 or position == 22 or position == 36) {
        const next_railway: usize = switch (position) { 7 => 15, 22 => 25, else => 5 };
        const next_utility: usize = if (position == 22) 28 else 12;
        addRaw(next, 0, doubles, probability / 16.0);
        addRaw(next, 10, 0, probability / 16.0);
        addRaw(next, 11, doubles, probability / 16.0);
        addRaw(next, 24, doubles, probability / 16.0);
        addRaw(next, 39, doubles, probability / 16.0);
        addRaw(next, 5, doubles, probability / 16.0);
        addRaw(next, next_railway, doubles, probability * 2.0 / 16.0);
        addRaw(next, next_utility, doubles, probability / 16.0);
        addLanding(next, (position + 37) % squares, doubles, probability / 16.0); // Go back three squares.
        addRaw(next, position, doubles, probability * 6.0 / 16.0);
    } else {
        addRaw(next, position, doubles, probability);
    }
}

pub fn main() !void {
    var probability: [states]f64 = [_]f64{0.0} ** states;
    probability[0] = 1.0;

    for (0..300) |_| {
        var next: [states]f64 = [_]f64{0.0} ** states;
        for (probability, 0..) |state_probability, state| {
            if (state_probability == 0.0) continue;
            const position = state / 3;
            const previous_doubles = state % 3;
            for (1..5) |first_die| {
                for (1..5) |second_die| {
                    const roll_probability = state_probability / 16.0;
                    if (first_die == second_die and previous_doubles == 2) {
                        addRaw(&next, 10, 0, roll_probability);
                    } else {
                        const doubles = if (first_die == second_die) previous_doubles + 1 else 0;
                        addLanding(&next, (position + first_die + second_die) % squares, doubles, roll_probability);
                    }
                }
            }
        }
        probability = next;
    }

    var total: [squares]f64 = [_]f64{0.0} ** squares;
    for (probability, 0..) |state_probability, state| total[state / 3] += state_probability;
    var first: usize = 0;
    for (1..squares) |square| {
        if (total[square] > total[first]) first = square;
    }
    var second: usize = if (first == 0) 1 else 0;
    for (0..squares) |square| {
        if (square != first and total[square] > total[second]) second = square;
    }
    var third: usize = if (first != 0 and second != 0) 0 else 1;
    for (0..squares) |square| {
        if (square != first and square != second and total[square] > total[third]) third = square;
    }

    var buffer: [64]u8 = undefined;
    var writer = std.Io.File.stdout().writer(std.Options.debug_io, &buffer);
    try writer.interface.print("{d:0>2}{d:0>2}{d:0>2}\n", .{ first, second, third });
    try writer.interface.flush();
}
